File: ros2-basic/topic_simple/src/node_helloworld_py/node_helloworld_py/node_publisher_by_helloworld_interface.py
import rclpy
from rclpy.node import Node
from node_helloworld_interface.msg import Message


class NodeHelloworld(Node):
    def __init__(self, name: str):
        super().__init__(name)
        self.count_ = 0
        # 创建 1 秒周期的定时器
        self.timer_ = self.create_timer(1.0, self.timer_callback)

        self.publisher_ = self.create_publisher(
            Message, "helloworld_py_by_helloworld_interface", 10
        )

    def timer_callback(self):
        msg = Message()
        msg.name = "publisher"
        msg.message = (
            f"Hello World Python(node_helloworld_interface) Publisher:  {self.count_}"
        )
        self.publisher_.publish(msg)
        self.get_logger().info(f"Hello World Python: {self.count_}")
        self.count_ += 1


def main(args=None):
    rclpy.init(args=args)
    node = NodeHelloworld("node_topic_publisher_helloworld_py_by_helloworld_interface")
    rclpy.spin(node)
    rclpy.shutdown()

File: ros2-basic/topic_simple/src/node_helloworld_cpp/src/node_publisher_by_helloworld_interface.cpp
#include "rclcpp/rclcpp.hpp"
#include "node_helloworld_interface/msg/message.hpp"

class NodeHelloworld : public rclcpp::Node
{
public:
    NodeHelloworld(std::string name) : Node(name)
    {
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&NodeHelloworld::timer_callback, this));

        publisher_ = this->create_publisher<node_helloworld_interface::msg::Message>("helloworld_cpp_by_helloworld_interface", 10);
    }

private:
    void timer_callback()
    {
        node_helloworld_interface::msg::Message msg;
        msg.name = "publisher";
        msg.message = "Hello World C++(node_helloworld_interface) Publisher: " + std::to_string(count_);
        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Hello World Topic C++(node_helloworld_interface) Publisher: %d", ++count_);
    }

private:
    int count_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<node_helloworld_interface::msg::Message>::SharedPtr publisher_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NodeHelloworld>("node_topic_publisher_helloworld_cpp_by_helloworld_interface");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}